pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1136 @@
+		},
+		Spec: networkingv1alpha1.ClusterCIDRSpec{
+			PerNodeHostBits: minPerNodeHostBits,
+		},
+	}
+
+	ipv4PerNodeHostBits := int32(math.MinInt32)
+	ipv6PerNodeHostBits := int32(math.MinInt32)
+	isDualstack := false
+	if len(allocatorParams.ClusterCIDRs) == 2 {
+		isDualstack = true
+	}
+
+	for i, cidr := range allocatorParams.ClusterCIDRs {
+		if netutil.IsIPv4CIDR(cidr) {
+			defaultCIDRConfig.Spec.IPv4 = cidr.String()
+			ipv4PerNodeHostBits = ipv4MaxCIDRMask - int32(allocatorParams.NodeCIDRMaskSizes[i])
+			if !isDualstack && ipv4PerNodeHostBits > minPerNodeHostBits {
+				defaultCIDRConfig.Spec.PerNodeHostBits = ipv4PerNodeHostBits
+			}
+		} else if netutil.IsIPv6CIDR(cidr) {
+			defaultCIDRConfig.Spec.IPv6 = cidr.String()
+			ipv6PerNodeHostBits = ipv6MaxCIDRMask - int32(allocatorParams.NodeCIDRMaskSizes[i])
+			if !isDualstack && ipv6PerNodeHostBits > minPerNodeHostBits {
+				defaultCIDRConfig.Spec.PerNodeHostBits = ipv6PerNodeHostBits
+			}
+		}
+	}
+
+	if isDualstack {
+		// In case of dualstack CIDRs, currently the default values for PerNodeMaskSize are
+		// 24 for IPv4 (PerNodeHostBits=8) and 64 for IPv6(PerNodeHostBits=64), there is no
+		// requirement for the PerNodeHostBits to be equal for IPv4 and IPv6, However with
+		// the introduction of ClusterCIDRs, we enforce the requirement for a single
+		// PerNodeHostBits field, thus we choose the minimum PerNodeHostBits value, to avoid
+		// overflow for IPv4 CIDRs.
+		if ipv4PerNodeHostBits >= minPerNodeHostBits && ipv4PerNodeHostBits <= ipv6PerNodeHostBits {
+			defaultCIDRConfig.Spec.PerNodeHostBits = ipv4PerNodeHostBits
+		} else if ipv6PerNodeHostBits >= minPerNodeHostBits && ipv6PerNodeHostBits <= ipv4MaxCIDRMask {
+			defaultCIDRConfig.Spec.PerNodeHostBits = ipv6PerNodeHostBits
+		}
+	}
+
+	existingConfigList.Items = append(existingConfigList.Items, *defaultCIDRConfig)
+
+	return
+}
+
+// reconcileCreate handles create ClusterCIDR events.
+func (r *multiCIDRRangeAllocator) reconcileCreate(clusterCIDR *networkingv1alpha1.ClusterCIDR) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if ccutil.NeedToAddFinalizer(clusterCIDR, clusterCIDRFinalizer) {
+		klog.V(3).Infof("Creating ClusterCIDR %s", clusterCIDR.Name)
+		if err := r.createClusterCIDR(clusterCIDR, false); err != nil {
+			klog.Errorf("Unable to create ClusterCIDR %s : %v", clusterCIDR.Name, err)
+			return err
+		}
+	}
+	return nil
+}
+
+// reconcileBootstrap handles creation of existing ClusterCIDRs.
+// adds a finalizer if not already present.
+func (r *multiCIDRRangeAllocator) reconcileBootstrap(clusterCIDR *networkingv1alpha1.ClusterCIDR) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	terminating := false
+	// Create the ClusterCIDR only if the Spec has not been modified.
+	if clusterCIDR.Generation > 1 {
+		terminating = true
+		err := fmt.Errorf("CIDRs from ClusterCIDR %s will not be used for allocation as it was modified", clusterCIDR.Name)
+		klog.Errorf("ClusterCIDR Modified: %v", err)
```

Comment:
should we exit here and return the error?

Reply:
ah, ok, it only needs to modify the termination var

Reply:
Yeah this is required so that we do not allocate from this CIDR in the future